### PR TITLE
have one validate function for the whole form

### PR DESCRIPTION
**Pull request by wochinge**
**Proposed changes**:
- fix https://github.com/RasaHQ/rasa/issues/6016

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by wochinge** [comment]
@Ghostvv I thought of something like this for custom slot mappings. 

I still have to work on making the extracted slot candidates clearly distinguishable for the custom `validate` function.

**Comment by Ghostvv** [comment]
let's merge [this one](https://github.com/RasaHQ/rasa/pull/6172) into here. Could you add a test that validation is called on the trigger input even if there is different activation action in between

File: rasa/core/actions/forms.py [hunk 1]
```diff
@@ -295,36 +294,56 @@ async def validate_slots(
...
+        _tracker = self._temporary_tracker(tracker, events, domain)
+        _action = RemoteAction(validate_name, self.action_endpoint)
+        validate_events = await _action.run(output_channel, nlg, _tracker, domain)
+
+        validated_slot_names = [
+            event.key for event in validate_events if isinstance(event, SlotSet)
+        ]
+
+        return validate_events + [
+            event for event in events if event.key not in validated_slot_names
```
Comment (Ghostvv) [inline rasa/core/actions/forms.py:332]: please add comment why we're doing this

File: rasa/core/actions/forms.py [hunk 2]
```diff
@@ -370,22 +389,50 @@ async def request_next_slot(
...
+            (
+                slot
+                for slot in self.required_slots(domain)
+                if self._should_request_slot(tracker, slot)
+            ),
+            None,
+        )
+
     def _name_of_utterance(self, domain: Domain, slot_name: Text) -> Text:
         full_name = f"utter_ask_{self._form_name}_{slot_name}"
```
Comment (Ghostvv) [inline rasa/core/actions/forms.py:437]: if the name starts with `utter_` it is not a custom action, should we first look for `full_name = f"action_ask_{self._form_name}_{slot_name}"` and only `run` in this case?

**Comment by Ghostvv** [comment]
> I still have to work on making the extracted slot candidates clearly distinguishable for the custom validate function.

this has to be part of the sdk, right?

**Comment by wochinge** [comment]
> I still have to work on making the extracted slot candidates clearly distinguishable for the custom validate function.

Yes 👍 However, I added a `ActionExecuted(form_action)` to the temporary tracker which we send to the SDK so that we can simply take all `SlotSet` events after the `ActionExecuted(form_action)` event

File: rasa/core/actions/forms.py (hunk 2 above)
Comment (wochinge) [reply rasa/core/actions/forms.py:436]: Let's do this after the alpha and see if users actually want a custom action to ask for the next slot.

Comment (Ghostvv) [reply rasa/core/actions/forms.py:436]: ok, let's remove `.run` there and add `TODO`

Comment (wochinge) [reply rasa/core/actions/forms.py:436]: The run is fine, what speaks against it? It will use `ActionUtterTemplate` and that's it?

Comment (Ghostvv) [reply rasa/core/actions/forms.py:436]: it just a bit counter intuitive (form how usually we call templates in custom actions), but I think it is actually fine

**Pull request merged**